pds4indextools/pds4_create_xml_index.py

```diff
@@ -428,41 +742,132 @@ def main():
         for url in xml_urls:
             update_nillable_elements_from_xsd_file(url, nillable_elements_info)
 
-        filepath = file.relative_to(args.directorypath)
+        filepath = str(file.relative_to(args.directorypath)).replace('\\', '/')
 
         namespaces = root.nsmap
         namespaces['pds'] = namespaces.pop(None)
         prefixes = {v: k for k, v in namespaces.items()}
 
-        xml_results = {}
-        traverse_and_store(root, tree, xml_results, elements_to_scrape,
+        label_results = {}
+        traverse_and_store(root, tree, label_results,
                            nillable_elements_info, config, file)
 
-        for key in list(xml_results.keys()):
-            process_tags(xml_results, key, root,
-                         namespaces, prefixes, args)
+        for key in list(label_results.keys()):
+            process_headers(label_results, key, root, namespaces, prefixes)
+
+        for key in list(label_results.keys()):
+            key_new = key.replace('[', '<')
+            key_new = key_new.replace(']', '>')
+            label_results[key_new] = label_results.pop(key)
+
+        for key in list(label_results.keys()):
+            parts = key.split('/')
+            new_parts = []
+            for part in parts:
+                if not part.endswith('>') and parts.index(part) != 1:
+                    part = part+'<1>'
+                    new_parts.append(part)
+                else:
+                    new_parts.append(part)
+            key_new = '/'.join(new_parts[1:])
+            label_results[key_new] = label_results.pop(key)
+
+        for key in list(label_results.keys()):
+            if 'cyfunction' in key:
+                del label_results[key]
+
+        xpath_map = renumber_xpaths(label_results.keys())
+        for old_xpath, new_xpath in xpath_map.items():
+            label_results[new_xpath] = label_results.pop(old_xpath)
+
+        correct_duplicates(label_results)
 
-        lid = xml_results.get('pds:logical_identifier', 'Missing_LID')
+        verboseprint(
+            'Now filtering label results according to given element file.')
+        label_results = filter_dict_by_glob_patterns(
+            label_results, elements_to_scrape, verboseprint)
+
+        if args.simplify_xpaths:
```

Comment:
MAJOR BUG HERE. This can't be done on a per-label basis. It has to be done for the entire collection of labels once at the end. For example, consider the elements file:
```
*Science_Facets*
```
and the command line:
```
python pds4_create_xml_index.py /mnt/rms-holdings/pds4-holdings/bundles/uranus_occs_earthbased/uranus_occ_u0_kao_91cm "**/rings/*.xml" "**/global/*.xml"  --output-file x.txt --dump-available-xpaths  --elements-file elements.txt
```
The resulting xpath file is:
```
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:wavelength_range<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:discipline_name<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:facet1<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:wavelength_range<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:discipline_name<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:facet1<1>
```
but now look what happens when I add `--simply-xpaths` to it:
```
pds:wavelength_range
pds:discipline_name
pds:facet1
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:wavelength_range<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:discipline_name<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<1>/pds:facet1<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:wavelength_range<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:discipline_name<1>
pds:Product_Observational/pds:Observation_Area<1>/pds:Primary_Result_Summary<1>/pds:Science_Facets<2>/pds:facet1<1>
```
We have some xpaths simplified and some not simplified.

You should check other things that you're doing on a per-label basis to make sure they don't need to be done after all the labels are read as well.

Reply:
This is something that should have been caught by the test suite, so make sure you add new tests to cover this kind of thing.

Reply:
Thank you for catching this. The new code works with your example, and all tests now pass.